packages/next-swc/crates/next-core/src/next_client_reference/ecmascript_client_reference/ecmascript_client_reference_proxy_module.rs

```diff
@@ -65,41 +66,72 @@ impl EcmascriptClientReferenceProxyModule {
     }
 
     #[turbo_tasks::function]
-    async fn proxy_module(self: Vc<Self>) -> Result<Vc<EcmascriptModuleAsset>> {
-        let this = self.await?;
+    async fn proxy_module(&self) -> Result<Vc<EcmascriptModuleAsset>> {
         let mut code = CodeBuilder::default();
 
-        // Adapted from
-        // next.js/packages/next/src/build/webpack/loaders/next-flight-loader/index.ts
+        let server_module_path = &*self.server_module_ident.path().to_string().await?;
+
         writedoc!(
             code,
             r#"
-                import {{ createProxy }} from 'next/dist/build/webpack/loaders/next-flight-loader/module-proxy'
+                import {{ createProxy }} from 'next/dist/build/webpack/loaders/next-flight-loader/module-proxy';
 
-                const proxy = createProxy({server_module_path})
+                const proxy = createProxy({server_module_path});
 
                 // Accessing the __esModule property and exporting $$typeof are required here.
                 // The __esModule getter forces the proxy target to create the default export
                 // and the $$typeof value is for rendering logic to determine if the module
                 // is a client boundary.
                 const {{ __esModule, $$typeof }} = proxy;
 
-                export {{ __esModule, $$typeof }};
                 export default proxy;
             "#,
-            server_module_path = StringifyJs(&this.server_module_ident.path().to_string().await?)
+            server_module_path = StringifyJs(server_module_path),
         )?;
 
+        // Adapted from
+        // next.js/packages/next/src/build/webpack/loaders/next-flight-loader/index.ts
+        if let EcmascriptExports::EsmExports(exports) = &*self.client_module.get_exports().await? {
+            let mut exported_names = exports
+                .await?
+                .exports
+                .keys()
+                .cloned()
+                .collect::<IndexSet<_>>();
+
+            for result in exports.expand_star_exports().await? {
+                let export_info = result.await?;
+
+                exported_names.extend(export_info.star_exports.iter().cloned());
+
+                if export_info.has_dynamic_exports {
+                    // TODO: throw? warn?
+                }
+            }
+
+            for (i, client_ref) in exported_names.into_iter().enumerate() {
+                writedoc!(
+                    code,
+                    r#"
+                        // Initialize export.
+                        const _e{cnt} = proxy[{client_ref}];
```

Comment:
At least add a comment that this getter triggers a side effect in the proxy implementation, but I would prefer it to be more explicit.